### PR TITLE
bugfix/issue-72: SMAP_RSS_L2_SSS_V4 subsets not finishing

### PR DESCRIPTION
**Pull request by sliu008**
Github Issue: #72 

### Description

SMAP_RSS_L2_SSS_V4 subsetting takes a long time. 

### Overview of work done

Update chunk function to be auto if we return an empty dictionary.

### Overview of verification done

Tested locally via command line and also tested with harmony l2ss container images

### Overview of integration done

There is also an increase in time to take to subset with smaller granules with auto chunking.

Screen Shot 2022-04-05 at 9 53 23 AM

Still took about 4 minutes for harmony call to finish when running locally.

Screen Shot 2022-04-05 at 9 42 51 AM

## PR checklist:

* [ ] Linted
* [ ] Updated unit tests
* [ ] Updated changelog
* [ ] Integration testing

_See [Pull Request Review Checklist](../CONTRIBUTING.md#reviewing) for pointers on reviewing this pull request_

### CODE REVIEW

File: podaac/subsetter/subset.py [hunk 1]
```diff
@@ -256,14 +256,21 @@ def calculate_chunks(dataset):
...
-    chunk_dict = {dim: 4000 for dim in dataset.dims
-                  if dataset.dims[dim] > 4000
-                  and len(dataset.dims) > 1}
-    return chunk_dict
+    chunk = {dim: 4000 for dim in dataset.dims
+             if dataset.dims[dim] > 4000
+             and len(dataset.dims) > 1}
+
+    if not chunk:
+        dataset_dtypes_hasobject = [dataset.variables[var].dtype.hasobject for var in dataset.variables]
```
Comment (skorper) [inline podaac/subsetter/subset.py:269]: What's the purpose of this line?

Comment (sliu008) [reply podaac/subsetter/subset.py:269]: > What's the purpose of this line?

check all variables dtype.hasobject which is a bool, and if any are true we can't use auto chunk as would throw this exception.

https://github.com/dask/dask/blob/main/dask/array/core.py#L3004

File: podaac/subsetter/subset.py [hunk 2]
```diff
@@ -819,8 +826,12 @@ def in_shape(lon, lat):
         point_in_shapefile = shapefile_df.contains(point)
         return point_in_shapefile.array[0]
 
+    dask = "forbidden"
+    if chunks:
+        dask = "allowed"
+
     in_shape_vec = np.vectorize(in_shape)
-    boolean_mask = xr.apply_ufunc(in_shape_vec, dataset[lon_var_name], dataset[lat_var_name])
+    boolean_mask = xr.apply_ufunc(in_shape_vec, dataset[lon_var_name], dataset[lat_var_name], dask=dask)
```
Comment (skorper) [inline podaac/subsetter/subset.py:834]: Also curious about this line -- if there are no chunks (or auto) won't the data be numpy arrays (vs dask arrays) and dask won't be used anyways? I'm curious about why this is needed

**Comment by skorper** [comment]
I'm wondering if there is some lower bound we can set to not use `auto` chunking. It seems like those have nearly doubled in processing time

**Comment by skorper** [comment]
> Still took about 4 minutes for harmony call to finish when running locally.

Wow that's quite long. Maybe we can try playing with the chunk sizes manually (without `auto`) and see how this changes? These granules are very large but maybe we can improve on 4 minutes

File: podaac/subsetter/subset.py (hunk 2 above)
Comment (sliu008) [reply podaac/subsetter/subset.py:834]: > Also curious about this line -- if there are no chunks (or auto) won't the data be numpy arrays (vs dask arrays) and dask won't be used anyways? I'm curious about why this is needed

When we use the chunk function a dask array is returned https://xarray.pydata.org/en/stable/generated/xarray.DataArray.chunk.html so if chunk is anything but {} we need to pass in dask=allowed in xr.apply_ufunc

**Comment by sliu008** [comment]
> I'm wondering if there is some lower bound we can set to not use `auto` chunking. It seems like those have nearly doubled in processing time

I can improve it by chunk on how many dimensions and that seem to improve the benchmark overall as the SMAP_RSS_L2_SSS_V4 had 6 dimensions. But didn't seem to matter what my chunk size were i didn't see any improvements on that 4 minute.

**Comment by sliu008** [comment]
Updated calculate chunk function based on dimensions improve benchmarks.

Screen Shot 2022-04-07 at 11 33 16 AM

**Comment by skorper** [comment]
The only major difference I see in the benchmarks is for `JA1_GPN_2PeP001_002_20020115_060706_20020115_070316.nc. I'm wondering why there's a change at all, because that granule appears to have only two dims, so should be using the older logic?

**Comment by sliu008** [comment]
> The only major difference I see in the benchmarks is for `JA1_GPN_2PeP001_002_20020115_060706_20020115_070316.nc. I'm wondering why there's a change at all, because that granule appears to have only two dims, so should be using the older logic?

Yes it should be using old logic

**Comment by skorper** [comment]
Looks good @sliu008. Thanks for this fix!!

**Review by skorper — APPROVED** [review]

**Comment by skorper** [comment]
This collection has been disabled in production until this issue is deployed to production. It should be reassociated once this goes to production.

[RSS SMAP Level 2C Sea Surface Salinity V4.0 Validated Dataset](https://mmt.earthdata.nasa.gov/collections/C2036880739-POCLOUD)

**Pull request merged**